### PR TITLE
PRO-2393 server side of PRO-2392, support for imposing custom login requirements

### PR DESCRIPTION
**Pull request by boutell**
Did this as TDD (Test Driven Development for anyone peeping in), feel good about the correctness as a result.

### CODE REVIEW

File: test/login-requirements.js [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    page = await apos.http.get(
+      '/',
+      {
+        jar
+      }
+    );
+
+    assert(page.match(/logged out/));
+
+
```
Comment (abea) [inline test/login-requirements.js:259]: eslint error here

File: modules/@apostrophecms/login/index.js [hunk 2]
```diff
@@ -106,34 +107,118 @@ module.exports = {
...
+          const passportLogin = (user) => {
+            return require('util').promisify(function(user, callback) {
+              return req.login(user, callback);
+            })(user);
+          };
+          // Don't make verify functions worry about whether this object
+          // is present, just the value of their own sub-property
+          req.body.requirements = req.body.requirements || {};
+
+          if (req.body.incompleteToken) {
```
Comment (abea) [inline modules/@apostrophecms/login/index.js:122]: The conditional flow would be easier to follow with the logic broken into separate functions.

**Review by abea — CHANGES_REQUESTED** [review]
Part of the tech design is:

> TheAposLogin should support configuration of custom Vue components to be brought into play in the login application.

That's followed by some additional detail about the browser experience. I don't see a separate ticket for those hooks. What is the plan for that? Am I misunderstanding the tech design? I don't see guidance for that layer.

File: modules/@apostrophecms/login/index.js (hunk 2 above)
Comment (boutell) [reply modules/@apostrophecms/login/index.js:122]: Agreed, I will refactor this route. See slack for a response about the tech design and tickets etc. This ticket is purely the server side.

**Comment by boutell** [comment]
Refactored so the degree of complexity per method is similar to that of the original route.

**Review by abea — APPROVED** [review]

**Pull request merged**